### ISSUE TITLE
On the running results and usage issues of the model

### ISSUE THREAD

**Issue by yuddecho** [Zhihan1996/DNABERT_2#27]
Hi, 

Thank you so much for DNABERT-2! There were some issues when I ran the model and used the results.

First, when I run the sample code and find hidden_ States. shape is torch. Size ([1, 17, 768]) not torch. Size ([1, len (DNA), 768])

```python
import torch
from transformers import AutoTokenizer, AutoModel

tokenizer = AutoTokenizer.from_pretrained("zhihan1996/DNABERT-2-117M", trust_remote_code=True)
model = AutoModel.from_pretrained("zhihan1996/DNABERT-2-117M", trust_remote_code=True)

dna = "ACGTAGCATCGGATCTATCTATCGACACTTGGTTATCGATCTACGAGCATCTCGTTAGC"
inputs = tokenizer(dna, return_tensors = 'pt')["input_ids"]
hidden_states = model(inputs)[0] # [1, sequence_length, 768]

# Question 1 : torch.Size([1, 17, 768]) not torch.Size([1, len(dna), 768])
print(hidden_states.shape)
```

Secondly, I have observed that the model (inputs) returns two tensors with sizes of torch. Size ([1, 17, 768]) and torch. Size ([1, 768]). I would like to ask what the second tensor means?

```python
import torch
from transformers import AutoTokenizer, AutoModel

tokenizer = AutoTokenizer.from_pretrained("zhihan1996/DNABERT-2-117M", trust_remote_code=True)
model = AutoModel.from_pretrained("zhihan1996/DNABERT-2-117M", trust_remote_code=True)

dna = "ACGTAGCATCGGATCTATCTATCGACACTTGGTTATCGATCTACGAGCATCTCGTTAGC"
inputs = tokenizer(dna, return_tensors='pt')["input_ids"]

result = model(inputs)

# output: 2
print(len(result))

# Question 2: What is the meaning of result[1]. shape
# output: torch.Size([1, 17, 768]) torch.Size([1, 768])
print(result[0].shape, result[1].shape)
```

Third, if multiple DNA sequences are inputted simultaneously and padding is set, how can I obtain the tensor encoding of the corresponding length for each sequence (instead of the length after padding)

```python
import torch
from transformers import AutoTokenizer, AutoModel

tokenizer = AutoTokenizer.from_pretrained("zhihan1996/DNABERT-2-117M", trust_remote_code=True)
model = AutoModel.from_pretrained("zhihan1996/DNABERT-2-117M", trust_remote_code=True)

dna = ["ACGTAGCATCGGATCTATTTAGC", "ACGTAGCATCGGATCTATCTATCGACACCTATCATCTCGTTAGC", "ACGTATTATCGATCTACGAGCATCTCGTTAGC"]
inputs = tokenizer(dna, return_tensors='pt', padding=True)["input_ids"]
hidden_states = model(inputs)[0]  # [1, sequence_length, 768]

# torch.Size([3, 13, 768])
print(hidden_states.shape)
```

For the third question, they did this when using [ESM](https://github.com/facebookresearch/esm) before. I want to obtain the same sequence_ representations.

```python
import torch

# Load ESM-2 model
model, alphabet = torch.hub.load("facebookresearch/esm:main", "esm2_t33_650M_UR50D")
batch_converter = alphabet.get_batch_converter()
model.eval()  # disables dropout for deterministic results

# Prepare data (first 2 sequences from ESMStructuralSplitDataset superfamily / 4)
data = [
    ("protein1", "MKTVRQERLKSIVRILERSKEPVSGAQLAEELSVSRQVIVQDIAYLRSLGYNIVATPRGYVLAGG"),
    ("protein2", "KALTARQQEVFDLIRDHISQTGMPPTRAEIAQRLGFRSPNAAEEHLKALARKGVIEIVSGASRGIRLLQEE"),
    ("protein2 with mask", "KALTARQQEVFDLIRD<mask>ISQTGMPPTRAEIAQRLGFRSPNAAEEHLKALARKGVIEIVSGASRGIRLLQEE"),
    ("protein3", "K A <mask> I S Q"),
]
batch_labels, batch_strs, batch_tokens = batch_converter(data)
batch_lens = (batch_tokens != alphabet.padding_idx).sum(1)

# Extract per-residue representations (on CPU)
with torch.no_grad():
    results = model(batch_tokens, repr_layers=[33], return_contacts=True)
token_representations = results["representations"][33]

# Output: torch.Size([4, 73, 1280])
print(token_representations.shape)

# Generate per-sequence representations via averaging
# NOTE: token 0 is always a beginning-of-sequence token, so the first residue is token 1.
sequence_representations = []
for i, tokens_len in enumerate(batch_lens):
    sequence_representation = token_representations[i, 1: tokens_len - 1].mean(0)

    # Output: torch.Size([1280])
    print(sequence_representation.shape)

    sequence_representations.append(sequence_representation)
```

Thanks !!!

**Comment by Zhihan1996** [comment]
Hey,

Thanks for your interest in this work.

1. this is expected since we use BPE for tokenization, which means the DNA sequence of length 10 may be tokenized into 3 tokens (e.g., `ACGTAGCATC` -> `ACGTA  GCA  TC`). And the 17 there is the length of the tokenized sequence.

2. The [1, 17, 768] one is the embedding of each token, while the [1, 768] is the pooled embedding of the [CLS] token. Since the [CLS] token is not trained during pre-training, we do not suggest using its embedding as sequence embedding.

3. I think you can use a similar approach as the code piece you shared to achieve this.
